GitHubReferences.ps1

```diff
@@ -0,0 +1,442 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+function Get-GitHubReference
+{
+<#
+    .SYNOPSIS
+        Retrieve a reference of a given GitHub repository.
+
+    .DESCRIPTION
+        Retrieve a reference of a given GitHub repository.
+        The Git repo for this module can be found here: http://aka.ms/PowerShellForGitHub
+
+    .PARAMETER OwnerName
+        Owner of the repository.
+        If not supplied here, the DefaultOwnerName configuration property value will be used.
+
+    .PARAMETER RepositoryName
+        Name of the repository.
+        If not supplied here, the DefaultRepositoryName configuration property value will be used.
+
+    .PARAMETER Uri
+        Uri for the repository.
+        The OwnerName and RepositoryName will be extracted from here instead of needing to provide
+        them individually.
+
+    .PARAMETER Reference
+        Name of the reference, for example: "heads/<branch name>" for branches and "tags/<tag name>" for tags.
+        Gets all the references (everything in the namespace, not only heads and tags) if not provided
+
+    .PARAMETER AccessToken
+        If provided, this will be used as the AccessToken for authentication with the
+        REST Api. Otherwise, will attempt to use the configured value or will run unauthenticated.
+
+    .PARAMETER NoStatus
+        If this switch is specified, long-running commands will run on the main thread
+        with no commandline status update.  When not specified, those commands run in
+        the background, enabling the command prompt to provide status information.
+        If not supplied here, the DefaultNoStatus configuration property value will be used.
+
+    .OUTPUTS
+        [PSCustomObject]
+        Details of the git reference in the given repository
+
+    .EXAMPLE
+        Get-GitHubReference -OwnerName microsoft -RepositoryName PowerShellForGitHub -Reference heads/master
+#>
+    [CmdletBinding(
+        SupportsShouldProcess,
+        DefaultParametersetName='Elements')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSShouldProcess", "", Justification="Methods called within here make use of PSShouldProcess, and the switch is passed on to them inherently.")]
+    param(
+        [Parameter(ParameterSetName='Elements')]
+        [string] $OwnerName,
+
+        [Parameter(ParameterSetName='Elements')]
+        [string] $RepositoryName,
+
+        [Parameter(
+            Mandatory,
+            ParameterSetName='Uri')]
+        [string] $Uri,
+
+        [string] $Reference,

```

Comment:
As I've thought about this more, I don't think it makes sense for users to have to know the syntax `heads/<branch name>` or `tags/<tag name>`.  Instead, I think you should split this out and have an optional parameter of `[string] $Tag` and another one as `[string] $Head`.  Users can just specify the one that should be used. You can defined them to be mutually exclusive via parametersets.